cdap-security/src/main/java/co/cask/cdap/security/authorization/RemoteAuthorizationEnforcer.java

```diff
@@ -0,0 +1,193 @@
+
+/**
+ * Remote implementation of the AuthorizationEnforcer. Contacts master for authorization enforcement and
+ * then caches the results if caching is enabled.
+ */
+public class RemoteAuthorizationEnforcer extends AbstractAuthorizationEnforcer {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RemoteAuthorizationEnforcer.class);
+
+  private static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(EntityId.class, new EntityIdTypeAdapter())
+    .create();
+
+  private final RemoteClient remoteClient;
+  private final boolean cacheEnabled;
+
+  private static LoadingCache<AuthorizationPrivilege, Boolean> authPolicyCache;
+
+  @Inject
+  public RemoteAuthorizationEnforcer(CConfiguration cConf, final DiscoveryServiceClient discoveryClient) {
+    super(cConf);
+    this.remoteClient = new RemoteClient(discoveryClient, Constants.Service.APP_FABRIC_HTTP,
+                                         new DefaultHttpRequestConfig(false), "/v1/execute/");
+    this.cacheEnabled = cConf.getBoolean(Constants.Security.Authorization.CACHE_ENABLED);
+    int cacheTtlSecs = cConf.getInt(Constants.Security.Authorization.CACHE_TTL_SECS);
+    int cacheMaxEntries = cConf.getInt(Constants.Security.Authorization.CACHE_MAX_ENTRIES);
+
+    authPolicyCache = CacheBuilder.newBuilder()
+      .expireAfterWrite(cacheTtlSecs, TimeUnit.SECONDS)
+      .maximumSize(cacheMaxEntries)
+      .build(new CacheLoader<AuthorizationPrivilege, Boolean>() {
+        @SuppressWarnings("NullableProblems")
+        @Override
+        public Boolean load(AuthorizationPrivilege authorizationPrivilege) throws Exception {
+          LOG.trace("Cache miss for " + authorizationPrivilege);
+          return doEnforce(authorizationPrivilege);
+        }
+      });
+  }
+
+  @Override
+  public void enforce(EntityId entity, Principal principal, Action action) throws Exception {
+    if (!isSecurityAuthorizationEnabled()) {
+      return;
+    }
+    AuthorizationPrivilege authorizationPrivilege = new AuthorizationPrivilege(principal, entity, action);
+
+    boolean allowed = cacheEnabled ? authPolicyCache.get(authorizationPrivilege) : doEnforce(authorizationPrivilege);
+    if (!allowed) {
+      throw new UnauthorizedException(principal, action, entity);
+    }
+  }
+
+  private boolean doEnforce(AuthorizationPrivilege authorizationPrivilege) throws IOException {
+    HttpResponse response = executeRequest(authorizationPrivilege);
+    return HttpResponseStatus.OK.getCode() == response.getResponseCode();
+
+  }
+
+  private HttpResponse executeRequest(AuthorizationPrivilege authorizationPrivilege) throws IOException {
+    HttpRequest request = remoteClient.requestBuilder(HttpMethod.POST, "enforce")
+      .withBody(GSON.toJson(createBody(authorizationPrivilege.getEntityId(),
+                                       authorizationPrivilege.getPrincipal(),
+                                       authorizationPrivilege.getAction())))
```

Comment:
In earlier comment I raised the need for having endpoint handling Set<Action> so that we make one call. It seems like AuthorizationPrivilege only have action. Also if I remember correctly so far we havent seen an enforce call with actions rather than action so might not be even needed.